shark_turbine/kernel/wave/register_analysis.py

```diff
@@ -33,16 +42,43 @@ def set_register_shape(trace: CapturedTrace, custom: CustomOp) -> None:
                     idx
                 ]
             )
-            set_register_shape(trace, iter_arg)
+            set_register_shape(trace, iter_arg, vector_map)
             custom.fx_node.thread_shape = iter_arg.fx_node.thread_shape
             break
+        elif isinstance(custom_user, ReduceOp):
+            get_thread_shape = lambda reg_shape: max(reg_shape)
+            # Check that dim is non-reduction && in hw_constraint.vector_shape.
+            is_valid_dim = lambda dim: dim != custom_user.dim and dim in vector_map
+            # TODO: Modify num_reduction_dims once we add support for multi-dim reduction.
+            num_reduction_dims = 1
+            register_shape = [
+                vector_map[dim]
+                for dim in custom_user.type.symbolic_shape
+                if is_valid_dim(dim)
+            ]
+            expected_result_rank = (
+                len(custom_user.type.symbolic_shape) - num_reduction_dims
+            )
+            # If rank do not match => some dims not found in hw_constraint.vector_shape.
+            if len(register_shape) != expected_result_rank:
+                raise NotImplementedError(
+                    "NYI: Handling of dim not in vector_shapes during register analysis."
+                )
+            num_thread_dims = sum(1 for dim in register_shape if dim > 1)
```

Comment:
Is this counting the number of non unit dims? If so can we call it that?

Reply:
Good idea, done!